numpy/lib/function_base.py

```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+    next_indexes = next_indexes.astype(np.intp)
+    return previous_indexes, next_indexes
+
+
+def _quantile(
+        arr: np.array,
+        quantiles: np.array,
+        axis: int = -1,
+        interpolation=QuantileInterpolation.INCLUSIVE,
+        out=None,
+):
+    """
+    Private function that doesn't support extended axis or keepdims.
+    These methods are extended to this function using _ureduce
+    See nanpercentile for parameter usage
+    It computes the quantiles of the array for the given axis.
+    A linear interpolation is performed based on the interpolation_type.
+
+    By default, the interpolation is QuantileInterpolation.INCLUSIVE where
+    alpha == beta == 1 which performs the 7th method of Hyndman&Fan.
+    With QuantileInterpolation.MEDIAN_UNBIASED we get  alpha == beta == 1/3
+    thus the 8th method of Hyndman&Fan.
+    """
+    # --- Setup
+    arr = np.asanyarray(arr)
+    values_count = arr.shape[axis]
+    # The dimensions of `q` are prepended to the output shape, so we need the
+    # axis being sampled from `arr` to be last.
+    DATA_AXIS = 0
+    arr = np.moveaxis(arr, axis, destination=DATA_AXIS)
+    # --- Computation of indexes
+    # Index where to find the value in the sorted array.
+    # Virtual because it is a floating point value, not an valid index.
+    # The nearest neighbours are used for interpolation
+    virtual_indexes = interpolation.virtual_index_fun(values_count, quantiles)
+    virtual_indexes = np.asanyarray(virtual_indexes)
+    previous_indexes, next_indexes = _get_indexes(arr,
+                                                  virtual_indexes,
+                                                  values_count)
+    # --- Sorting
+    arr.partition(
+        np.unique(np.concatenate(([0, -1],
+                                  previous_indexes.ravel(),
+                                  next_indexes.ravel(),
+                                  ))),
+        axis=DATA_AXIS)
+    if np.issubdtype(arr.dtype, np.inexact):
+        slices_having_nans = np.isnan(take(arr, indices=-1, axis=DATA_AXIS))
+    else:
+        slices_having_nans = None
+    # --- Get values from indexes
+    previous = np.take(arr, previous_indexes, axis=DATA_AXIS)
+    next = np.take(arr, next_indexes, axis=DATA_AXIS)
+    # --- Linear interpolation
+    gamma = _get_gamma(arr.dtype,
+                       virtual_indexes,
+                       previous_indexes,
+                       interpolation)
+    result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
+    gamma = gamma.reshape(result_shape)
+    result = _linear_interpolation_formula(previous,
+                                           next,
+                                           gamma,
+                                           out=out)
+    if np.any(slices_having_nans):
+        if result.ndim == 0 and out is None:
             # can't write to a scalar
-            r = a.dtype.type(np.nan)
+            result = np.asanyarray(np.nan)
```

Comment:
```suggestion
            result = np.array(np.nan, dtype=result.dtype)
```
I guess this needs some other type, such as `a.dtype` here.  Returning `np.nan` seems incorrect, since it will always be double precision.  `result.dtype` may not work for `object` dtype, but I guess that case is wrong anyway, since it does not handle NaNs anyways (also before these changes).

Reply:
A test using `float128` as input may be good.

Reply:
I added a unit test for this using `AllFloat` and your suggestion works better than what I wrote but I still have a few cases going wrong.
It goes:
from `float16` to `float64`; 
from `float32` to `float64`
and from `float64` to `float128`

I will try to fix this as well.

Reply:
I fixed it using `arr.dtype` instead of `result.dtype`. It should be safe thanks to `if np.any(slices_having_nans):` above.